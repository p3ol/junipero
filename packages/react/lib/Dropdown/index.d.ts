import { Placement } from '@floating-ui/react-dom-interactions';
import {
  Props as clickProps,
} from '@floating-ui/react-dom-interactions/src/hooks/useClick';
import {
  Props as dismissProps,
} from '@floating-ui/react-dom-interactions/src/hooks/useDismiss';
import {
  Props as hoverProps,
} from '@floating-ui/react-dom-interactions/src/hooks/useHover';
import React, { MutableRefObject } from 'react';

export declare type DropdownRef = {
  opened: Boolean;
  toggle: () => void;
  open: () => void;
  close: () => void;
  isJunipero: Boolean;
  innerRef: MutableRefObject<any>;
};

declare interface DropdownProps extends React.ComponentPropsWithRef<any> {
  className?: String;
  container?: String | React.ReactNode;
  floatingOptions?: Object;
  clickOptions?: clickProps;
  hoverOptions?: hoverProps;
  dismissOptions?: dismissProps;
  openend?: Boolean;
  placement?: Placement;
  trigger?: 'click' | 'hover' | 'manual';
  onToggle?: (props: { opened: Boolean }) => void;
  disabled?: Boolean;
  ref?: MutableRefObject<DropdownRef | undefined>;
}

declare function Dropdown(props: DropdownProps): JSX.Element;

export default Dropdown;
