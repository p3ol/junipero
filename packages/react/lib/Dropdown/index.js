import {
  forwardRef,
  useCallback,
  useImperativeHandle,
  useRef,
  useReducer,
  useEffect,
} from 'react';
import { classNames, mockState } from '@junipero/core';
import {
  offset,
  autoUpdate,
  flip,
  shift,
  safePolygon,
  useFloating,
  useInteractions,
  useDismiss,
  useClick,
  useHover,
} from '@floating-ui/react-dom-interactions';
import PropTypes from 'prop-types';

import { DropdownContext } from '../contexts';

const Dropdown = forwardRef(({
  className,
  container,
  disabled,
  floatingOptions,
  clickOptions,
  hoverOptions,
  dismissOptions,
  opened = false,
  placement = 'bottom-start',
  trigger = 'click',
  onToggle,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const [state, dispatch] = useReducer(mockState, {
    opened,
  });
  const { x, y, reference, floating, strategy, context } = useFloating({
    open: state.opened,
    onOpenChange: (...args) => onOpenChange(...args),
    placement,
    strategy: 'fixed',
    whileElementsMounted: autoUpdate,
    ...floatingOptions || {},
    middleware: floatingOptions?.middleware || [
      offset(10),
      flip(),
      shift(),
    ],
  });
  const { getReferenceProps, getFloatingProps } = useInteractions([
    useClick(context, {
      enabled: trigger === 'click',
      ...clickOptions || {},
    }),
    useHover(context, {
      enabled: trigger === 'hover',
      handleClose: safePolygon(),
      ...hoverOptions || {},
    }),
    useDismiss(context, {
      enabled: trigger === 'click',
      ...dismissOptions || {},
    }),
  ]);

  useEffect(() => {
    if (disabled && state.opened) {
      close();
    }
  }, [disabled]);

  useImperativeHandle(ref, () => ({
    opened: state.opened,
    toggle,
    open,
    close,
    isJunipero: true,
    innerRef,
  }));

  const onOpenChange = o => {
    if (disabled && o) {
      return;
    }

    dispatch({ opened: o });
    onToggle?.({ opened: o });
  };

  const toggle = () => {
    if (disabled) {
      return;
    }

    if (state.opened) {
      close();
    } else {
      open();
    }
  };

  const open = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: true });
    onToggle?.({ opened: true });
  };

  const close = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: false });
    onToggle?.({ opened: false });
  };

  const getContext = useCallback(() => ({
    opened: state.opened,
    container,
    x,
    y,
    reference,
    floating,
    strategy,
    toggle,
    open,
    close,
    getReferenceProps,
    getFloatingProps,
  }), [
    state.opened,
    x,
    y,
    reference,
    floating,
    strategy,
  ]);

  return (
    <DropdownContext.Provider value={getContext()}>
      <div
        { ...rest }
        className={classNames(
          'junipero dropdown',
          {
            opened: state.opened,
            disabled,
          },
          className
        )}
        ref={innerRef}
      />
    </DropdownContext.Provider>
  );
});

Dropdown.displayName = 'Dropdown';
Dropdown.propTypes = {
  clickOptions: PropTypes.object,
  container: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.object,
    PropTypes.node,
  ]),
  disabled: PropTypes.bool,
  dismissOptions: PropTypes.object,
  floatingOptions: PropTypes.object,
  hoverOptions: PropTypes.object,
  opened: PropTypes.bool,
  placement: PropTypes.string,
  trigger: PropTypes.oneOf(['click', 'hover', 'manual']),
  onToggle: PropTypes.func,
};

export default Dropdown;
