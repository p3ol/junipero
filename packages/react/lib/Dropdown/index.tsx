import {
  type RefObject,
  type ReactElement,
  useCallback,
  useImperativeHandle,
  useRef,
  useReducer,
  useEffect,
} from 'react';
import { classNames, omit, mockState } from '@junipero/core';
import {
  type UseClickProps,
  type UseDismissProps,
  type UseHoverProps,
  type Placement,
  type OpenChangeReason,
  type UseFloatingOptions,
  type Boundary,
  offset,
  autoUpdate,
  flip,
  shift,
  safePolygon,
  useFloating,
  useInteractions,
  useDismiss,
  useClick,
  useHover,
} from '@floating-ui/react';

import type { JuniperoRef, SpecialComponentPropsWithRef } from '../types';
import { DropdownContext, type DropdownContextType } from '../contexts';
import AccessibilityStore from '../AccessibilityStore';

export declare interface DropdownRef extends JuniperoRef {
  opened: boolean;
  toggle(): void;
  open(): void;
  close(): void;
  innerRef: RefObject<HTMLDivElement>;
}

export declare interface DropdownProps extends Omit<
  SpecialComponentPropsWithRef<'div', DropdownRef>,
  'onToggle'
>{
  clickOptions?: UseClickProps;
  container?: string | ReactElement | DocumentFragment | HTMLElement;
  disabled?: boolean;
  dismissOptions?: UseDismissProps;
  floatingOptions?: UseFloatingOptions & {
    boundary?: Boundary;
  };
  hoverOptions?: UseHoverProps;
  opened?: boolean;
  placement?: Placement;
  trigger?: 'click' | 'hover' | 'manual';
  withAccessibility?: boolean;
  onToggle?(props: { opened: boolean }): void;
}

export declare interface DropdownState {
  opened: boolean;
  visible: boolean;
  highlightedOptionId: string | null;
}

const Dropdown = ({
  ref,
  className,
  container,
  disabled,
  floatingOptions,
  clickOptions,
  hoverOptions,
  dismissOptions,
  opened = false,
  placement = 'bottom-start',
  trigger = 'click',
  withAccessibility = true,
  onToggle,
  ...rest
}: DropdownProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const [state, dispatch] = useReducer(mockState<DropdownState>, {
    opened: opened ?? false,
    visible: opened ?? false,
    highlightedOptionId: null,
  });
  const { x, y, refs, strategy, context } = useFloating({
    open: state.opened,
    onOpenChange: (...args) => onOpenChange(...args),
    placement,
    strategy: 'fixed',
    whileElementsMounted: autoUpdate,
    ...omit(floatingOptions || {}, ['boundary', 'middleware']),
    middleware: floatingOptions?.middleware || [
      offset(10),
      flip({
        boundary: floatingOptions?.boundary ||
          floatingOptions?.elements?.reference,
      }),
      shift({
        boundary: floatingOptions?.boundary ||
          floatingOptions?.elements?.reference,
      }),
    ],
  });
  const { getReferenceProps, getFloatingProps } = useInteractions([
    useClick(context, {
      enabled: trigger === 'click',
      ...clickOptions || {},
    }),
    useHover(context, {
      enabled: trigger === 'hover',
      handleClose: safePolygon(),
      ...hoverOptions || {},
    }),
    useDismiss(context, {
      enabled: trigger === 'click',
      ...dismissOptions || {},
    }),
  ]);

  useEffect(() => {
    if (disabled && state.opened) {
      close();
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [disabled]);

  useImperativeHandle(ref, () => ({
    opened: state.opened,
    toggle,
    open,
    close,
    isJunipero: true,
    innerRef,
  }));

  const onOpenChange = (o: boolean, _: Event, __: OpenChangeReason) => {
    if (disabled && o) {
      return;
    }

    dispatch({ opened: o, visible: true });
    onToggle?.({ opened: o });
  };

  const open = useCallback(() => {
    if (disabled) {
      return;
    }

    dispatch({ opened: true, visible: true });
    onToggle?.({ opened: true });
  }, [disabled, onToggle]);

  const close = useCallback(() => {
    if (disabled) {
      return;
    }

    dispatch({ opened: false });
    onToggle?.({ opened: false });
  }, [disabled, onToggle]);

  const toggle = useCallback(() => {
    if (disabled) {
      return;
    }

    if (state.opened) {
      close();
    } else {
      open();
    }
  }, [disabled, state.opened, open, close]);

  const onAnimationExit = useCallback(() => {
    dispatch({ visible: false });
  }, []);

  const getContext = useCallback<() => DropdownContextType>(() => ({
    opened: state.opened,
    visible: state.visible,
    container,
    x,
    y,
    refs,
    strategy,
    toggle,
    open,
    close,
    getReferenceProps,
    getFloatingProps,
    onAnimationExit,
  }), [
    state.opened, state.visible,
    x, y, refs, strategy,
    toggle, open, close,
    getReferenceProps, getFloatingProps, onAnimationExit, container,
  ]);

  const dropdown = (
    <DropdownContext.Provider value={getContext()}>
      <div
        { ...rest }
        className={classNames(
          'junipero dropdown',
          {
            opened: state.opened,
            disabled,
          },
          className
        )}
        ref={innerRef}
      />
    </DropdownContext.Provider>
  );

  return withAccessibility ? (
    <AccessibilityStore>{ dropdown }</AccessibilityStore>
  ) : dropdown;
};

Dropdown.displayName = 'Dropdown';

export default Dropdown;
