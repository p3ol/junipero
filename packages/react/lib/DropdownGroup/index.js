import { classNames } from '@junipero/core';
import PropTypes from 'prop-types';

const DropdownGroup = ({ children, title, className }) => (
  <div className={classNames('dropdown-group', className)}>
    <div className="junipero extra group-title">{ title }</div>
    <div className="options">
      { children }
    </div>
  </div>
);

DropdownGroup.displayName = 'DropdownGroup';
DropdownGroup.propTypes = {
  title: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.node,
    PropTypes.func,
  ]),
};

export default DropdownGroup;
