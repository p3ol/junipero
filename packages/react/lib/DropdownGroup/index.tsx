import type { ComponentPropsWithoutRef, ReactNode } from 'react';
import { classNames } from '@junipero/core';
import PropTypes from 'prop-types';

export declare interface DropdownGroupProps
  extends ComponentPropsWithoutRef<any> {
  children?: JSX.Element | ReactNode;
  className?: string;
  title?: ReactNode | JSX.Element;
}

const DropdownGroup = ({
  children,
  title,
  className,
}: DropdownGroupProps): JSX.Element => (
  <div className={classNames('dropdown-group', className)}>
    <div className="junipero extra group-title">{ title }</div>
    <div className="options">
      { children }
    </div>
  </div>
);

DropdownGroup.displayName = 'DropdownGroup';
DropdownGroup.propTypes = {
  title: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.node,
    PropTypes.func,
  ]),
};

export default DropdownGroup;
