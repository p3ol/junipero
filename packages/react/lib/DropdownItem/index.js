import { classNames } from '@junipero/core';

const DropdownItem = ({ className, ...rest }) => (
  <li className={classNames('dropdown-item', className)} {...rest} />
);

DropdownItem.displayName = 'DropdownItem';

export default DropdownItem;
