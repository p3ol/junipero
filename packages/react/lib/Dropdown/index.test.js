import { act, fireEvent, render, renderHook } from '@testing-library/react';
import { slideInDownMenu } from '@junipero/transitions';

import { useDropdown } from '../hooks';
import DropdownToggle from '../DropdownToggle';
import DropdownMenu from '../DropdownMenu';
import DropdownGroup from '../DropdownGroup';
import DropdownItem from '../DropdownItem';
import Dropdown from './index';

describe('<Dropdown />', () => {
  it('should render', () => {
    const { getByText, container, unmount } = render(
      <Dropdown>
        <DropdownToggle><span>Click me</span></DropdownToggle>
        <DropdownMenu>
          <DropdownItem>Item 1</DropdownItem>
        </DropdownMenu>
      </Dropdown>
    );
    expect(container).toMatchSnapshot();

    fireEvent.click(getByText('Click me'));

    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should auto close when disabled prop changes', () => {
    const { container, unmount, rerender, getByText } = render(
      <Dropdown opened={true}>
        <DropdownToggle><span>Click me</span></DropdownToggle>
        <DropdownMenu>
          <DropdownItem>Item 1</DropdownItem>
        </DropdownMenu>
      </Dropdown>
    );

    expect(container).toMatchSnapshot();

    rerender(
      <Dropdown disabled>
        <DropdownToggle><span>Click me</span></DropdownToggle>
        <DropdownMenu>
          <DropdownItem>Item 1</DropdownItem>
        </DropdownMenu>
      </Dropdown>
    );

    fireEvent.click(getByText('Click me'));

    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should allow to render menu inside another element', () => {
    const { container, unmount } = render(
      <>
        <div id="menu-container" />
        <Dropdown container="#menu-container" opened={true}>
          <DropdownToggle><span>Click me</span></DropdownToggle>
          <DropdownMenu>
            <DropdownItem>Item 1</DropdownItem>
          </DropdownMenu>
        </Dropdown>
      </>
    );

    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should allow to use groups', () => {
    const { container, unmount } = render(
      <Dropdown opened={true}>
        <DropdownToggle><span>Click me</span></DropdownToggle>
        <DropdownMenu>
          <DropdownGroup title="Group 1">
            <DropdownItem>Item 1</DropdownItem>
            <DropdownItem>Item 2</DropdownItem>
          </DropdownGroup>
          <DropdownGroup title="Group 2">
            <DropdownItem>Item 3</DropdownItem>
            <DropdownItem>Item 4</DropdownItem>
          </DropdownGroup>
        </DropdownMenu>
      </Dropdown>
    );

    expect(container).toMatchSnapshot();

    unmount();
  });
});

describe('useDropdwon()', () => {
  it('should render hook', () => {
    const { result } = renderHook(() => useDropdown(), { wrapper: Dropdown });
    expect(
      Object.keys(result.current)
    ).toEqual(expect.arrayContaining(
      [
        'opened',
        'x',
        'y',
        'reference',
        'floating',
        'strategy',
        'toggle',
        'open',
        'close',
        'getReferenceProps',
        'getFloatingProps',
      ]
    ));
  });
  it('should set hook with props', () => {
    const container = <div />;

    const wrapper = ({ children }) => {

      return (
        <Dropdown
          container={container}
          opened={true}
        >
          {children}
        </Dropdown>
      );
    };

    const { result } = renderHook(() => useDropdown(), { wrapper });
    expect(result.current.container).toEqual(container);
    expect(result.current.opened).toBeTruthy();
  });

  it('should open and close dropdown', () => {
    const toggleMock = jest.fn();

    const wrapper = ({ children }) => {

      return (
        <Dropdown
          onToggle={toggleMock}
        >
          {children}
        </Dropdown>
      );
    };

    const { result } = renderHook(() => useDropdown(), { wrapper });
    expect(result.current.opened).toBeFalsy();

    act(() => result.current.open());
    expect(result.current.opened).toBeTruthy();
    expect(toggleMock).toHaveBeenCalledWith({ opened: true });

    act(() => result.current.close());
    expect(result.current.opened).toBeFalsy();
    expect(toggleMock).toHaveBeenCalledWith({ opened: false });

    act(() => result.current.toggle());
    expect(result.current.opened).toBeTruthy();
    expect(toggleMock).toHaveBeenCalledWith({ opened: true });

    act(() => result.current.toggle());
    expect(result.current.opened).toBeFalsy();
    expect(toggleMock).toHaveBeenCalledWith({ opened: false });
  });

  it('should not open and close dropdown if disabled', () => {
    const toggleMock = jest.fn();

    const wrapper = ({ children }) => {

      return (
        <Dropdown
          onToggle={toggleMock}
          disabled={true}
        >
          {children}
        </Dropdown>
      );
    };

    const { result } = renderHook(() => useDropdown(), { wrapper });
    expect(result.current.opened).toBeFalsy();

    act(() => result.current.open());
    expect(result.current.opened).toBeFalsy();
    expect(toggleMock).not.toHaveBeenCalled();
  });

  it('should allow to animate dropdown', () => {
    jest.useFakeTimers();
    const { getByText, container, unmount } = render(
      <Dropdown>
        <DropdownToggle><span>Click me</span></DropdownToggle>
        <DropdownMenu animate={slideInDownMenu}>
          <DropdownItem>Item 1</DropdownItem>
        </DropdownMenu>
      </Dropdown>
    );
    expect(container).toMatchSnapshot('Enter: Not in dom');

    // Enter
    fireEvent.click(getByText('Click me'));
    act(() => { jest.advanceTimersByTime(100); });
    expect(container).toMatchSnapshot('Enter: In dom');

    // Exit
    fireEvent.click(getByText('Click me'));
    act(() => { jest.advanceTimersByTime(100); });
    expect(container).toMatchSnapshot('Exit: Not in dom');

    jest.useRealTimers();
    unmount();
  });

  it('should keep the menu in dom when animated and apparition is css', () => {
    jest.useFakeTimers();
    const { getByText, container, unmount } = render(
      <Dropdown>
        <DropdownToggle><span>Click me</span></DropdownToggle>
        <DropdownMenu animate={slideInDownMenu} apparition="css">
          <DropdownItem>Item 1</DropdownItem>
        </DropdownMenu>
      </Dropdown>
    );
    expect(container).toMatchSnapshot('Enter: Already in dom');

    // Enter
    fireEvent.click(getByText('Click me'));
    act(() => { jest.advanceTimersByTime(100); });
    expect(container).toMatchSnapshot('Enter: Still in dom');

    // Exit
    fireEvent.click(getByText('Click me'));
    act(() => { jest.advanceTimersByTime(100); });
    expect(container).toMatchSnapshot('Exit: Still in dom');

    jest.useRealTimers();
    unmount();
  });
});
