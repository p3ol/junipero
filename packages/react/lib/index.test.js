import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
