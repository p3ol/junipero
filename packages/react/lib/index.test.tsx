import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
