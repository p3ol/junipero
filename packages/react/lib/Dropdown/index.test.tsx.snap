// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Dropdown /> should allow to render menu inside another element 1`] = `
<div>
  <div
    id="menu-container"
  >
    <div
      class="junipero dropdown-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-«rb»-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <li
          class="junipero dropdown-item"
          id="junipero-dropdown-item-«rf»"
          tabindex="0"
        >
          Item 1
        </li>
      </ul>
    </div>
  </div>
  <div
    class="junipero dropdown opened"
    id="junipero-dropdown-«rb»"
  >
    <span
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-«rb»-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-«rc»"
      role="combobox"
    >
      Click me
    </span>
  </div>
</div>
`;

exports[`<Dropdown /> should allow to use groups 1`] = `
<div>
  <div
    class="junipero dropdown opened"
    id="junipero-dropdown-«rh»"
  >
    <span
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-«rh»-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-«ri»"
      role="combobox"
    >
      Click me
    </span>
    <div
      class="junipero dropdown-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-«rh»-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <div
          class="dropdown-group"
        >
          <div
            class="junipero extra group-title"
          >
            Group 1
          </div>
          <div
            class="options"
          >
            <li
              class="junipero dropdown-item"
              id="junipero-dropdown-item-«rk»"
              tabindex="0"
            >
              Item 1
            </li>
            <li
              class="junipero dropdown-item"
              id="junipero-dropdown-item-«rl»"
              tabindex="0"
            >
              Item 2
            </li>
          </div>
        </div>
        <div
          class="dropdown-group"
        >
          <div
            class="junipero extra group-title"
          >
            Group 2
          </div>
          <div
            class="options"
          >
            <li
              class="junipero dropdown-item"
              id="junipero-dropdown-item-«rm»"
              tabindex="0"
            >
              Item 3
            </li>
            <li
              class="junipero dropdown-item"
              id="junipero-dropdown-item-«rn»"
              tabindex="0"
            >
              Item 4
            </li>
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<Dropdown /> should auto close when disabled prop changes 1`] = `
<div>
  <div
    class="junipero dropdown opened"
    id="junipero-dropdown-«r6»"
  >
    <span
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-«r6»-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-«r7»"
      role="combobox"
    >
      Click me
    </span>
    <div
      class="junipero dropdown-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-«r6»-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <li
          class="junipero dropdown-item"
          id="junipero-dropdown-item-«r9»"
          tabindex="0"
        >
          Item 1
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<Dropdown /> should auto close when disabled prop changes 2`] = `
<div>
  <div
    class="junipero dropdown disabled"
    id="junipero-dropdown-«r6»"
  >
    <span
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-«r6»-menu"
      aria-expanded="false"
      aria-haspopup="listbox"
      class="dropdown-toggle"
      dir="ltr"
      id="junipero-dropdown-toggle-«r7»"
      role="combobox"
    >
      Click me
    </span>
  </div>
</div>
`;

exports[`<Dropdown /> should render 1`] = `
<div>
  <div
    class="junipero dropdown"
    id="junipero-dropdown-«r1»"
  >
    <span
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-«r1»-menu"
      aria-expanded="false"
      aria-haspopup="listbox"
      class="dropdown-toggle"
      dir="ltr"
      id="junipero-dropdown-toggle-«r2»"
      role="combobox"
    >
      Click me
    </span>
  </div>
</div>
`;

exports[`<Dropdown /> should render 2`] = `
<div>
  <div
    class="junipero dropdown opened"
    id="junipero-dropdown-«r1»"
  >
    <span
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-«r1»-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-«r2»"
      role="combobox"
    >
      Click me
    </span>
    <div
      class="junipero dropdown-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-«r1»-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <li
          class="junipero dropdown-item"
          id="junipero-dropdown-item-«r4»"
          tabindex="0"
        >
          Item 1
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`useDropdwon() should allow to animate dropdown: Enter: In dom 1`] = `
<div>
  <div
    class="junipero dropdown opened"
    id="junipero-dropdown-«r11»"
  >
    <span
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-«r11»-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-«r12»"
      role="combobox"
    >
      Click me
    </span>
    <div
      class="junipero dropdown-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-«r11»-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner jp-slide-in-down-menu-enter-done"
      >
        <li
          class="junipero dropdown-item"
          id="junipero-dropdown-item-«r14»"
          tabindex="0"
        >
          Item 1
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`useDropdwon() should allow to animate dropdown: Enter: Not in dom 1`] = `
<div>
  <div
    class="junipero dropdown"
    id="junipero-dropdown-«r11»"
  >
    <span
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-«r11»-menu"
      aria-expanded="false"
      aria-haspopup="listbox"
      class="dropdown-toggle"
      dir="ltr"
      id="junipero-dropdown-toggle-«r12»"
      role="combobox"
    >
      Click me
    </span>
  </div>
</div>
`;

exports[`useDropdwon() should allow to animate dropdown: Exit: Not in dom 1`] = `
<div>
  <div
    class="junipero dropdown"
    id="junipero-dropdown-«r11»"
  >
    <span
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-«r11»-menu"
      aria-expanded="false"
      aria-haspopup="listbox"
      class="dropdown-toggle"
      dir="ltr"
      id="junipero-dropdown-toggle-«r12»"
      role="combobox"
    >
      Click me
    </span>
  </div>
</div>
`;

exports[`useDropdwon() should keep the menu in dom when animated and apparition is css: Enter: Already in dom 1`] = `
<div>
  <div
    class="junipero dropdown"
    id="junipero-dropdown-«r16»"
  >
    <span
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-«r16»-menu"
      aria-expanded="false"
      aria-haspopup="listbox"
      class="dropdown-toggle"
      dir="ltr"
      id="junipero-dropdown-toggle-«r17»"
      role="combobox"
    >
      Click me
    </span>
    <div
      class="junipero dropdown-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-«r16»-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    />
  </div>
</div>
`;

exports[`useDropdwon() should keep the menu in dom when animated and apparition is css: Enter: Still in dom 1`] = `
<div>
  <div
    class="junipero dropdown opened"
    id="junipero-dropdown-«r16»"
  >
    <span
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-«r16»-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-«r17»"
      role="combobox"
    >
      Click me
    </span>
    <div
      class="junipero dropdown-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-«r16»-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner jp-slide-in-down-menu-enter-done"
      >
        <li
          class="junipero dropdown-item"
          id="junipero-dropdown-item-«r19»"
          tabindex="0"
        >
          Item 1
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`useDropdwon() should keep the menu in dom when animated and apparition is css: Exit: Still in dom 1`] = `
<div>
  <div
    class="junipero dropdown"
    id="junipero-dropdown-«r16»"
  >
    <span
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-«r16»-menu"
      aria-expanded="false"
      aria-haspopup="listbox"
      class="dropdown-toggle"
      dir="ltr"
      id="junipero-dropdown-toggle-«r17»"
      role="combobox"
    >
      Click me
    </span>
    <div
      class="junipero dropdown-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-«r16»-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    />
  </div>
</div>
`;
